ChronoKeeper/StoryPipeline.cpp

```diff
@@ -1,44 +1,271 @@
+    	std::time_t time_t_chunk_end = std::chrono::high_resolution_clock::to_time_t(chunk_end_point); 
+    	std::cout <<"StoryPipeline: storyId { " << storyId<<"} prepend chunk {" << timelineStart<<" "<< std::ctime(&time_t_chunk_start) 
+		<<"} {" << timelineStart-chunkGranularity <<" " <<std::ctime(&time_t_chunk_end) <<"}"<<std::endl;
+#endif 
+	auto result= storyTimelineMap.insert( std::pair<uint64_t, chronolog::StoryChunk>
+			 ( timelineStart-chunkGranularity, chronolog::StoryChunk(storyId, timelineStart-chunkGranularity, timelineStart)));
+	if( !result.second)
+	{ 
+	    return storyTimelineMap.end(); 
+	}	
+        else
+	{
+	    timelineStart -= chunkGranularity;
+	    return result.first;
+	}
+}
+/////////////////////////////
+
+std::map<uint64_t, chronolog::StoryChunk>::iterator chronolog::StoryPipeline::appendStoryChunk()
+{
+  // append the next storyChunk at the end of storyTimeline and return the iterator to the new node
+#ifdef TRACE_CHUNKING
+    	std::chrono::time_point<std::chrono::system_clock,std::chrono::nanoseconds> epoch_time_point{};
+    	auto chunk_start_point = epoch_time_point + std::chrono::nanoseconds(timelineEnd);
+    	std::time_t time_t_chunk_start = std::chrono::high_resolution_clock::to_time_t(chunk_start_point);
+    	auto chunk_end_point = epoch_time_point + std::chrono::nanoseconds(timelineEnd+chunkGranularity);
+    	std::time_t time_t_chunk_end = std::chrono::high_resolution_clock::to_time_t(chunk_end_point); 
+    	std::cout <<"StoryPipeline: storyId { "<< storyId<<"} append chunk {" << timelineEnd <<" "<< std::ctime(&time_t_chunk_start) 
+		<<"} {" << timelineEnd+chunkGranularity <<" " <<std::ctime(&time_t_chunk_end) <<"}"<<std::endl;
+#endif 
+	auto result= storyTimelineMap.insert( std::pair<uint64_t, chronolog::StoryChunk>( timelineEnd, chronolog::StoryChunk(storyId, timelineEnd, timelineEnd+chunkGranularity)));
+	if( !result.second)
+	{ 
+	    return storyTimelineMap.end(); 
+	}	
+        else
+	{
+	    timelineEnd += chunkGranularity;
+	    return result.first;
+	}
+}
+//////////////////////
+
+void chronolog::StoryPipeline::collectIngestedEvents()
+{
+	activeIngestionHandle->swapActiveDeque(); 
+
+	if( !activeIngestionHandle->getPassiveDeque().empty())
+	{   mergeEvents(activeIngestionHandle->getPassiveDeque()); }
 
-return status;
 }
 ////////////////////
 
-int chronolog::StoryPipeline::mergeEvents(chronolog::StoryChunk const&)
+void chronolog::StoryPipeline::mergeEvents(chronolog::EventDeque & event_deque)
+{
+
+	if(event_deque.empty())
+	{ return;}
+
+	std::lock_guard<std::mutex> lock(sequencingMutex);
+	chl::LogEvent event;
+	// the last chunk is most likely the one that would get the events, so we'd start with the last 
+	// chunk and do the lookup only if it's not the one
+	// NOTE: we should never have less than 2 chunks in the active storyTimelineMap !!!
+    	std::map<uint64_t, chronolog::StoryChunk>::iterator chunk_to_merge_iter = --storyTimelineMap.end();
```

Comment:
I agree that most likely the last chunk will be the chunk to insert the events into. But I think we should still add the code to check the chunk lists in the storypipeline from newest to oldest to find out which chunk to insert when the event is in between <timelinestart, timelineend> but older than starttime of last chunk.

Reply:
what you are suggesting is done in the next line of execution : check for event.time() < last chunk.startTime is done first thing in insertEvent on line 164, and if this is the case than line 168 storyTimelineMAp.upper_bound(event.time()) find the older chunk the event would be inserted into